firedrake/preconditioners/fdm.py

```diff
@@ -1,128 +1,179 @@
-                y = I.createVecLeft()
-                x = I.createVecRight()
-                x.set(1.0E0)
-                I.mult(x, y)
-                vectors.append(y)
-                x.destroy()
-            return PETSc.NullSpace().create(constant=False, vectors=vectors, comm=nsp.getComm())
-
-        # Matrix-free assembly of the transformed Jacobian
+        e_fdm = element.reconstruct(variant=self._variant)
+
         if element == e_fdm:
             V_fdm, J_fdm, bcs_fdm = (V, J, bcs)
-            Amat, _ = pc.getOperators()
-            self._ctx_ref = octx
         else:
+            # Reconstruct Jacobian and bcs with variant element
             V_fdm = firedrake.FunctionSpace(V.mesh(), e_fdm)
-            J_fdm = ufl.replace(J, {t: t.reconstruct(function_space=V_fdm) for t in J.arguments()})
-            bcs_fdm = tuple(bc.reconstruct(V=V_fdm) for bc in bcs)
-            self.fdm_interp = prolongation_matrix_matfree(V, V_fdm, [], bcs_fdm)
-            self.A = allocate_matrix(J_fdm, bcs=bcs_fdm, form_compiler_parameters=fcp, mat_type=mat_type,
-                                     options_prefix=options_prefix)
-            self._assemble_A = partial(assemble, J_fdm, tensor=self.A, bcs=bcs_fdm,
-                                       form_compiler_parameters=fcp, mat_type=mat_type)
-            self._assemble_A()
-            Amat = self.A.petscmat
-
-            omat, _ = pc.getOperators()
-            inject = prolongation_matrix_matfree(V_fdm, V, [], [])
-            Amat.setNullSpace(interp_nullspace(inject, omat.getNullSpace()))
-            Amat.setTransposeNullSpace(interp_nullspace(inject, omat.getTransposeNullSpace()))
-            Amat.setNearNullSpace(interp_nullspace(inject, omat.getNearNullSpace()))
-            self.work_vec_x = Amat.createVecLeft()
-            self.work_vec_y = Amat.createVecRight()
-
+            J_fdm = J(*[t.reconstruct(function_space=V_fdm) for t in J.arguments()], coefficients={})
+            bcs_fdm = []
+            for bc in bcs:
+                W = V_fdm
+                for index in bc._indices:
+                    W = W.sub(index)
+                bcs_fdm.append(bc.reconstruct(V=W, g=0))
+
+            # Create a new _SNESContext in the variant space
             self._ctx_ref = self.new_snes_ctx(pc, J_fdm, bcs_fdm, mat_type,
                                               fcp=fcp, options_prefix=options_prefix)
 
-        if len(bcs) > 0:
-            self.bc_nodes = numpy.unique(numpy.concatenate([bcdofs(bc, ghost=False) for bc in bcs]))
-        else:
-            self.bc_nodes = numpy.empty(0, dtype=PETSc.IntType)
+            # Construct interpolation from variant to original spaces
+            self.fdm_interp = prolongation_matrix_matfree(V_fdm, V, bcs_fdm, [])
+            self.work_vec_x = Amat.createVecLeft()
+            self.work_vec_y = Amat.createVecRight()
+            if use_amat:
+                from firedrake.assemble import allocate_matrix, TwoFormAssembler
```

Comment:
Could these imports be at the top of the file? They should only be here to avoid cyclic import issues.

Reply:
This is required to avoid a cyclic import.

Reply:
This is required to avoid a cyclic import.